--- a/arch/mips/include/asm/ptrace.h
+++ b/arch/mips/include/asm/ptrace.h
@@ -138,7 +138,13 @@ extern int ptrace_set_watch_regs(struct task_struct *child,
 /*
  * Does the process account for user or for system time?
  */
+#if defined(CONFIG_MIPS_SF2000) && !defined(CONFIG_MMU)
+#define user_mode(regs) \
+	((((regs)->cp0_status & KU_MASK) == KU_USER) || \
+	 (((regs)->cp0_epc & 0xfc000000UL) == 0x84000000UL))
+#else
 #define user_mode(regs) (((regs)->cp0_status & KU_MASK) == KU_USER)
+#endif
 
 static inline int is_syscall_success(struct pt_regs *regs)
 {
--- a/arch/mips/kernel/entry.S
+++ b/arch/mips/kernel/entry.S
@@ -41,6 +41,15 @@ resume_userspace_check:
 	LONG_L	t0, PT_STATUS(sp)		# returning to kernel mode?
 	andi	t0, t0, KU_USER
+#if defined(CONFIG_MIPS_SF2000) && !defined(CONFIG_MMU)
+	bnez	t0, resume_userspace
+	LONG_L	t0, PT_EPC(sp)
+	li	t1, 0xfc000000
+	and	t0, t1
+	li	t1, 0x84000000
+	bne	t0, t1, resume_kernel
+#else
 	beqz	t0, resume_kernel
+#endif
 
 resume_userspace:
 	local_irq_disable			# make sure we dont miss an
@@ -206,6 +215,16 @@ FEXPORT(syscall_exit_partial)
 	LONG_L	t0, PT_STATUS(sp)		# returning to kernel mode?
 	andi	t0, t0, KU_USER
+#if defined(CONFIG_MIPS_SF2000) && !defined(CONFIG_MMU)
+	bnez	t0, 1f
+	LONG_L	t0, PT_EPC(sp)
+	li	t1, 0xfc000000
+	and	t0, t1
+	li	t1, 0x84000000
+	bne	t0, t1, resume_kernel
+1:
+#else
 	beqz	t0, resume_kernel
+#endif
 	li	t0, _TIF_WORK_SYSCALL_EXIT
 	and	t0, a2			# a2 is preloaded with TI_FLAGS
 	beqz	t0, work_pending	# trace bit set?
