--- a/drivers/usb/musb/sf2000.c
+++ b/drivers/usb/musb/sf2000.c
@@ -18,10 +18,78 @@ struct sf2000_glue {
 	struct platform_device *usb_phy;
 };
 
+static struct musb_fifo_cfg sf2000_fifo_cfg[] = {
+	{ .hw_ep_num = 1, .style = FIFO_TX, .maxpacket = 512 },
+	{ .hw_ep_num = 1, .style = FIFO_RX, .maxpacket = 512 },
+	{ .hw_ep_num = 2, .style = FIFO_TX, .maxpacket = 512 },
+	{ .hw_ep_num = 2, .style = FIFO_RX, .maxpacket = 512 },
+	{ .hw_ep_num = 3, .style = FIFO_TX, .maxpacket = 512 },
+	{ .hw_ep_num = 3, .style = FIFO_RX, .maxpacket = 512 },
+	{ .hw_ep_num = 4, .style = FIFO_TX, .maxpacket = 512 },
+	{ .hw_ep_num = 4, .style = FIFO_RX, .maxpacket = 512 },
+	{ .hw_ep_num = 5, .style = FIFO_TX, .maxpacket = 512 },
+	{ .hw_ep_num = 5, .style = FIFO_RX, .maxpacket = 512 },
+};
+
+static const struct musb_hdrc_config sf2000_musb_config = {
+	.fifo_cfg = sf2000_fifo_cfg,
+	.fifo_cfg_size = ARRAY_SIZE(sf2000_fifo_cfg),
+	.multipoint = 1,
+	.dyn_fifo = 1,
+	.num_eps = 7,
+	.ram_bits = 12,
+};
+
+static irqreturn_t sf2000_musb_interrupt(int irq, void *data)
+{
+	struct musb *musb = data;
+	irqreturn_t ret = IRQ_NONE;
+	unsigned long flags;
+
+	spin_lock_irqsave(&musb->lock, flags);
+	musb->int_rx = musb_readw(musb->mregs, MUSB_INTRRX) &
+		musb_readw(musb->mregs, MUSB_INTRRXE);
+	musb->int_tx = musb_readw(musb->mregs, MUSB_INTRTX) &
+		musb_readw(musb->mregs, MUSB_INTRTXE);
+	musb->int_usb = musb_readb(musb->mregs, MUSB_INTRUSB) &
+		musb_readb(musb->mregs, MUSB_INTRUSBE);
+	if (musb->int_rx || musb->int_tx || musb->int_usb)
+		ret = musb_interrupt(musb);
+	spin_unlock_irqrestore(&musb->lock, flags);
+
+	return ret;
+}
+
+static int sf2000_musb_init(struct musb *musb)
+{
+	musb->xceiv = devm_usb_get_phy(musb->controller->parent,
+					 USB_PHY_TYPE_USB2);
+	if (IS_ERR(musb->xceiv))
+		return PTR_ERR(musb->xceiv);
+
+	musb->dyn_fifo = true;
+	musb->isr = sf2000_musb_interrupt;
+	return 0;
+}
+
+static int sf2000_musb_exit(struct musb *musb)
+{
+	return 0;
+}
+
+static const struct musb_platform_ops sf2000_musb_ops = {
+	.quirks = MUSB_DMA_INVENTRA | MUSB_INDEXED_EP,
+	.init = sf2000_musb_init,
+	.exit = sf2000_musb_exit,
+};
+
 static int sf2000_usb_probe(struct platform_device *pdev)
 {
 	const struct musb_hdrc_platform_data pdata = {
 		.mode = MUSB_HOST,
+		.power = 250,
+		.config = &sf2000_musb_config,
+		.platform_ops = &sf2000_musb_ops,
 	};
 	struct sf2000_glue *glue;
 	struct platform_device *musb;
@@ -50,6 +118,11 @@ static int sf2000_usb_probe(struct platform_device *pdev)
 	musb->dev.dma_mask = &musb->dev.coherent_dma_mask;
 	musb->dev.coherent_dma_mask = DMA_BIT_MASK(32);
 
+	ret = platform_device_add_resources(musb, pdev->resource,
+					    pdev->num_resources);
+	if (ret)
+		goto err_put;
+
 	ret = platform_device_add_data(musb, &pdata, sizeof(pdata));
 	if (ret)
 		goto err_put;
