--- a/Rules.mak
+++ b/Rules.mak
@@ -325,9 +325,11 @@
 PICFLAG-$(UCLIBC_FORMAT_FDPIC_ELF) := -mfdpic
 PICFLAG-$(UCLIBC_FORMAT_DSBT_ELF)  := -mdsbt -fpic
 PICFLAG := $(PICFLAG-y)
 ifeq ($(UCLIBC_FORMAT_FLAT),y)
+ifneq ($(STATIC_PIE),y)
 PICFLAG :=
 endif
+endif
 PIEFLAG_NAME:=-fPIE
 
 $(eval $(call check-gcc-var,-fdata-sections))
@@ -407,6 +409,7 @@
 
 ifeq ($(TARGET_ARCH),mips)
+	PICFLAG += -mabicalls
 	OPTIMIZATION+=-mno-split-addresses
 	CPU_CFLAGS-$(CONFIG_MIPS_N64_ABI)+=-mabi=64
 	CPU_CFLAGS-$(CONFIG_MIPS_O32_ABI)+=-mabi=32
