--- a/ldso/include/ldso.h
+++ b/ldso/include/ldso.h
@@ -48,7 +48,7 @@
 /* Pull in the MIN macro */
 #include <sys/param.h>
 /* Pull in the ldso syscalls and string functions */
-#if !defined(__ARCH_HAS_NO_SHARED__) || !defined(__ARCH_HAS_NO_LDSO__)
+#if !defined(__ARCH_HAS_NO_SHARED__) || !defined(__ARCH_HAS_NO_LDSO__) || defined(__STATIC_PIE__)
 #include <dl-syscall.h>
 #include <dl-string.h>
 /* Now the ldso specific headers */
--- a/ldso/include/dl-defs.h
+++ b/ldso/include/dl-defs.h
@@ -72,7 +72,7 @@
 #endif
 
 #ifdef _LIBC
-#ifndef __ARCH_HAS_NO_LDSO__
+#if !defined(__ARCH_HAS_NO_LDSO__) || defined(__STATIC_PIE__)
 /* arch specific defines */
 #include <dl-sysdep.h>
 #endif
